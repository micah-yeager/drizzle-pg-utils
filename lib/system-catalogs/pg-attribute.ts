import { boolean, char, integer, pgView, text } from "drizzle-orm/pg-core"
import { aclitem } from "./columns/aclitem"
import { anyarray } from "./columns/anyarray"
import { name } from "./columns/name"
import { oid } from "./columns/oid"

/** @see https://www.postgresql.org/docs/current/catalog-pg-attribute.html */
export const pgAttribute = pgView("pg_attribute", {
	attrelid: oid().notNull(),
	attname: name().notNull(),
	atttypid: oid().notNull(),
	attlen: integer().notNull(),
	attnum: integer().notNull(),
	attcacheoff: integer().notNull(),
	atttypmod: integer().notNull(),
	attndims: integer().notNull(),
	attbyval: boolean().notNull(),
	attalign: char().notNull(),
	attstorage: char().notNull(),
	attcompression: char().notNull(),
	attnotnull: boolean().notNull(),
	atthasdef: boolean().notNull(),
	atthasmissing: boolean().notNull(),
	attidentity: char().notNull(),
	attgenerated: char().notNull(),
	attisdropped: boolean().notNull(),
	attislocal: boolean().notNull(),
	attinhcount: integer().notNull(),
	attcollation: oid().notNull(),
	attstattarget: integer(),
	attacl: aclitem().array(),
	attoptions: text().array(),
	attfdwoptions: text().array(),
	attmissingval: anyarray(),
}).existing()
