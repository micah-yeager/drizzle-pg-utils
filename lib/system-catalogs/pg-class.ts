import { boolean, char, integer, pgView, real, text } from "drizzle-orm/pg-core"
import { aclitem } from "./columns/aclitem"
import { name } from "./columns/name"
import { oid } from "./columns/oid"
import { pgNodeTree } from "./columns/pg-node-tree"
import { xid } from "./columns/xid"

/** @see https://www.postgresql.org/docs/current/catalog-pg-class.html */
export const pgClass = pgView("pg_class", {
	oid: oid().notNull(),
	relname: name().notNull(),
	relnamespace: oid().notNull(),
	reltype: oid().notNull(),
	reloftype: oid().notNull(),
	relowner: oid().notNull(),
	relam: oid().notNull(),
	relfilenode: oid().notNull(),
	reltablespace: oid().notNull(),
	relpages: integer().notNull(),
	reltuples: real().notNull(),
	relallvisible: integer().notNull(),
	reltoastrelid: oid().notNull(),
	relhasindex: boolean().notNull(),
	relisshared: boolean().notNull(),
	relpersistence: char().notNull(),
	relkind: char().notNull(),
	relnatts: integer().notNull(),
	relchecks: integer().notNull(),
	relhasrules: boolean().notNull(),
	relhastriggers: boolean().notNull(),
	relhassubclass: boolean().notNull(),
	relrowsecurity: boolean().notNull(),
	relforcerowsecurity: boolean().notNull(),
	relispopulated: boolean().notNull(),
	relreplident: char().notNull(),
	relispartition: boolean().notNull(),
	relrewrite: oid().notNull(),
	relfrozenxid: xid().notNull(),
	relminmxid: xid().notNull(),
	relacl: aclitem().array(),
	reloptions: text().array(),
	relpartbound: pgNodeTree(),
}).existing()
