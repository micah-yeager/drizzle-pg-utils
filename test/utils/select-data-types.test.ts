import { expect, it } from "vitest"
import { pgClass } from "../../lib/system-catalogs/pg-class"
import { bpcharTable } from "../schema"
import { selectDataTypes } from "./select-data-types"

it("should allow table as input", async () => {
	await expect(selectDataTypes(bpcharTable)).resolves.toMatchSnapshot()
})

it("should allow view as input", async () => {
	await expect(selectDataTypes(pgClass)).resolves.toMatchSnapshot()
})
