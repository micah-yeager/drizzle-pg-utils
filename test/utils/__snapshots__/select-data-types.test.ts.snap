// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should allow table as input 1`] = `
[
  {
    "columnName": "id",
    "dataType": "integer",
  },
  {
    "columnName": "bpchar",
    "dataType": "bpchar",
  },
]
`;

exports[`should allow view as input 1`] = `
[
  {
    "columnName": "oid",
    "dataType": "oid",
  },
  {
    "columnName": "relname",
    "dataType": "name",
  },
  {
    "columnName": "relnamespace",
    "dataType": "oid",
  },
  {
    "columnName": "reltype",
    "dataType": "oid",
  },
  {
    "columnName": "reloftype",
    "dataType": "oid",
  },
  {
    "columnName": "relowner",
    "dataType": "oid",
  },
  {
    "columnName": "relam",
    "dataType": "oid",
  },
  {
    "columnName": "relfilenode",
    "dataType": "oid",
  },
  {
    "columnName": "reltablespace",
    "dataType": "oid",
  },
  {
    "columnName": "relpages",
    "dataType": "integer",
  },
  {
    "columnName": "reltuples",
    "dataType": "real",
  },
  {
    "columnName": "relallvisible",
    "dataType": "integer",
  },
  {
    "columnName": "reltoastrelid",
    "dataType": "oid",
  },
  {
    "columnName": "relhasindex",
    "dataType": "boolean",
  },
  {
    "columnName": "relisshared",
    "dataType": "boolean",
  },
  {
    "columnName": "relpersistence",
    "dataType": ""char"",
  },
  {
    "columnName": "relkind",
    "dataType": ""char"",
  },
  {
    "columnName": "relnatts",
    "dataType": "smallint",
  },
  {
    "columnName": "relchecks",
    "dataType": "smallint",
  },
  {
    "columnName": "relhasrules",
    "dataType": "boolean",
  },
  {
    "columnName": "relhastriggers",
    "dataType": "boolean",
  },
  {
    "columnName": "relhassubclass",
    "dataType": "boolean",
  },
  {
    "columnName": "relrowsecurity",
    "dataType": "boolean",
  },
  {
    "columnName": "relforcerowsecurity",
    "dataType": "boolean",
  },
  {
    "columnName": "relispopulated",
    "dataType": "boolean",
  },
  {
    "columnName": "relreplident",
    "dataType": ""char"",
  },
  {
    "columnName": "relispartition",
    "dataType": "boolean",
  },
  {
    "columnName": "relrewrite",
    "dataType": "oid",
  },
  {
    "columnName": "relfrozenxid",
    "dataType": "xid",
  },
  {
    "columnName": "relminmxid",
    "dataType": "xid",
  },
  {
    "columnName": "relacl",
    "dataType": "aclitem[]",
  },
  {
    "columnName": "reloptions",
    "dataType": "text[]",
  },
  {
    "columnName": "relpartbound",
    "dataType": "pg_node_tree",
  },
]
`;

exports[`should retrieve all columns 1`] = `
[
  {
    "columnName": "oid",
    "dataType": "oid",
  },
  {
    "columnName": "relname",
    "dataType": "name",
  },
  {
    "columnName": "relnamespace",
    "dataType": "oid",
  },
  {
    "columnName": "reltype",
    "dataType": "oid",
  },
  {
    "columnName": "reloftype",
    "dataType": "oid",
  },
  {
    "columnName": "relowner",
    "dataType": "oid",
  },
  {
    "columnName": "relam",
    "dataType": "oid",
  },
  {
    "columnName": "relfilenode",
    "dataType": "oid",
  },
  {
    "columnName": "reltablespace",
    "dataType": "oid",
  },
  {
    "columnName": "relpages",
    "dataType": "integer",
  },
  {
    "columnName": "reltuples",
    "dataType": "real",
  },
  {
    "columnName": "relallvisible",
    "dataType": "integer",
  },
  {
    "columnName": "reltoastrelid",
    "dataType": "oid",
  },
  {
    "columnName": "relhasindex",
    "dataType": "boolean",
  },
  {
    "columnName": "relisshared",
    "dataType": "boolean",
  },
  {
    "columnName": "relpersistence",
    "dataType": ""char"",
  },
  {
    "columnName": "relkind",
    "dataType": ""char"",
  },
  {
    "columnName": "relnatts",
    "dataType": "smallint",
  },
  {
    "columnName": "relchecks",
    "dataType": "smallint",
  },
  {
    "columnName": "relhasrules",
    "dataType": "boolean",
  },
  {
    "columnName": "relhastriggers",
    "dataType": "boolean",
  },
  {
    "columnName": "relhassubclass",
    "dataType": "boolean",
  },
  {
    "columnName": "relrowsecurity",
    "dataType": "boolean",
  },
  {
    "columnName": "relforcerowsecurity",
    "dataType": "boolean",
  },
  {
    "columnName": "relispopulated",
    "dataType": "boolean",
  },
  {
    "columnName": "relreplident",
    "dataType": ""char"",
  },
  {
    "columnName": "relispartition",
    "dataType": "boolean",
  },
  {
    "columnName": "relrewrite",
    "dataType": "oid",
  },
  {
    "columnName": "relfrozenxid",
    "dataType": "xid",
  },
  {
    "columnName": "relminmxid",
    "dataType": "xid",
  },
  {
    "columnName": "relacl",
    "dataType": "aclitem[]",
  },
  {
    "columnName": "reloptions",
    "dataType": "text[]",
  },
  {
    "columnName": "relpartbound",
    "dataType": "pg_node_tree",
  },
]
`;
